### ISSUE TITLE
04 05 06 Proyectos diabetes

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/benjaminmullermiranda-a11y-Machine_Learning#4]
### **Decision Tree (benjamin04-decision-tree-diabetes.ipynb)**

**Puntos fuertes:**
- EDA completo: histogramas, boxplots, heatmap de correlaciones PERO FALTAN INTERPRETACIONES DE GRÁFICOS
- Reemplazas 0s con mediana (enfoque correcto)
- Comparas **gini, entropy, log_loss** (los 3 criterios)
- Visualización de árbol con `plot_tree`
- Matrices de confusión para cada criterio
- Feature importance
- `stratify=y` en train_test_split
- Guardas el dataset limpio

**Falta:**
- **Comentarios interpretando los gráficos** aún siendo un EDA completo me falta algo más de interpretación con tus observaciones
- **No usas GridSearchCV** - solo comparas criterios manualmente sin optimizar otros hiperparámetros
- **No guardas el modelo** - solo guardas el CSV limpio, no el modelo entrenado

---

### **Random Forest (benajmin05-random-forest-diabetes.ipynb)**

**Puntos fuertes:**
- Análisis de impacto de `n_estimators` (5-300)
- Análisis de impacto de `max_depth` (1-20, None)
- Visualización de curvas de accuracy vs hiperparámetros
- Feature importance
- **Guardas el modelo** con joblib
- StandardScaler aplicado

**Falta:**
- **No usas GridSearchCV** - bucles manuales en lugar de búsqueda sistemática
- No usas `class_weight` para el desbalanceo
- Solo optimizas 2 hiperparámetros individualmente (no combinados)

---

### **Boosting (benjamin06-boosting-diabetes.ipynb)**

**Puntos fuertes:**
- Usas XGBoost
- Guardas el modelo con pickle
- Conclusión final comparando los 3 modelos

**Problemas críticos:**
- **No usa GridSearchCV** - solo pruebas valores fijos manualmente
- **No hay exploración de hiperparámetros** - usas directamente `n_estimators=100, learning_rate=0.1, max_depth=5`
- Código repetido (entrenas el mismo modelo dos veces)
- Variable `acc` usada sin definir previamente (error en la gráfica)
- Notebook muy corto y poco desarrollado
- No hay análisis del impacto de `scale_pos_weight` para el desbalanceo

---

### **Problemas principales:**

1. **No usas GridSearchCV en ningún notebook** - Este es un requisito explícito mencionado en las instrucciones ("Paso 4: Optimiza el modelo anterior... utilizando un grid search")

2. **Decision Tree no guardas el modelo** - Las instrucciones dicen "Paso 5: Guarda el modelo"

3. **Boosting muy superficial** - El notebook muy corto, sin exploración real de hiperparámetros

---

### **Código que falta - GridSearchCV para Decision Tree:**

```python
from sklearn.model_selection import GridSearchCV

param_grid = {
    'criterion': ['gini', 'entropy', 'log_loss'],
    'max_depth': [3, 5, 7, 10, None],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 5],
    'class_weight': ['balanced', None]
}

grid_search = GridSearchCV(
    DecisionTreeClassifier(random_state=42),
    param_grid,
    cv=5,
    scoring='recall'  # Priorizar recall en contexto médico
)
grid_search.fit(X_train, y_train)

print(f"Mejores parámetros: {grid_search.best_params_}")
best_model = grid_search.best_estimator_

# Guardar modelo
joblib.dump(best_model, './models/decision_tree_diabetes.pkl')
```

### **Código que falta - GridSearchCV para Boosting:**

```python
from sklearn.model_selection import GridSearchCV
from xgboost import XGBClassifier

param_grid = {
    'n_estimators': [50, 100, 200],
    'max_depth': [3, 5, 7],
    'learning_rate': [0.01, 0.1, 0.2],
    'scale_pos_weight': [1, 2, 3]  # Para desbalanceo de clases
}

grid_search = GridSearchCV(
    XGBClassifier(random_state=42, eval_metric='logloss'),
    param_grid,
    cv=5,
    scoring='recall'
)
grid_search.fit(X_train, y_train)

print(f"Mejores parámetros: {grid_search.best_params_}")
```

--- 

@benjaminmullermiranda-a11y 
**Conclusión:** Tu trabajo está muy bien lineas generales y muestra comprensión básica de los conceptos pero falta optimización sistemática requerida. El EDA del Decision Tree es bueno, pero Random Forest y especialmente Boosting necesitan más desarrollo.
